pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,308 @@
+			Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+			Expect(mapObj["spec"]).To(BeEquivalentTo(targetResource.Object["spec"]))
+			Expect(mapObj["metadata"]).To(BeEquivalentTo(targetResource.Object["metadata"]))
+		})
+
+		It("the returned resource's placement should be an empty matcher", func() {
+			// This test implicitly tests that the nested hierarchy of fields exists
+			mapObj, found, err := getMatchLabelsField(expectedTyped)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+			Expect(mapObj).To(BeEmpty())
+		})
+
+		It("the returned resource's namespace should be the input's namespace", func() {
+			Expect(expectedTyped.GetNamespace()).
+				To(Equal(resourceTyped.GetNamespace()))
+		})
+	})
+}
+
+// func assertNestedFields(accessor interface{}, expectedObj types.GomegaMatcher, obj map[string]interface{}, fields ...string) { }
+func testDistribute() {
+	var (
+		resourceTyped       *corev1.Pod
+		fedResource, fedPod *unstructured.Unstructured
+		f                   *federator
+		err                 error
+		initObjs            []runtime.Object
+		clusterNames        []string
+		scheme              *runtime.Scheme
+	)
+
+	BeforeEach(func() {
+		resourceTyped = newPod("test-pod")
+		scheme = runtime.NewScheme()
+	})
+
+	JustBeforeEach(func() {
+		f = newFederatorWithScheme(scheme, initObjs...)
+		err = f.Distribute(resourceTyped, clusterNames...)
+	})
+
+	When("the resource type scheme has NOT been added to the type registry", func() {
+		It("should throw an error", func() {
+			Expect(err).To(HaveOccurred())
+		})
+	})
+
+	When("the resource type scheme has been added to the type registry", func() {
+		BeforeEach(func() {
+			corev1.AddToScheme(scheme)
+			tmpFederator := newFederatorWithScheme(scheme)
+			fedResource, _ = tmpFederator.createFederatedResource(resourceTyped)
+		})
+
+		JustBeforeEach(func() {
+			fedPod = getResourceFromAPI(f.kubeFedClient, fedResource)
+		})
+
+		It("should return no error on success", func() {
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		When("the resource is already in the Kube API", func() {
+			expectedImage := "apache"
+
+			BeforeEach(func() {
+				initObjs = append(initObjs, fedResource)
+
+				resourceTyped.Spec = corev1.PodSpec{
+					Containers: []corev1.Container{
+						corev1.Container{
+							Image: expectedImage,
+							Name:  "httpd",
+						},
+					},
+				}
+			})
+
+			It("should update the resource", func() {
+				fedPodObj, _, _ := getTemplateField(fedPod)
+				expectObj, _, _ := getTemplateField(fedResource)
+
+				Expect(fedPodObj["spec"]).ToNot(BeEquivalentTo(expectObj["spec"]))
+
+				mapObj, _, _ := unstructured.NestedSlice(
+					fedPod.Object,
+					ctlutil.SpecField, ctlutil.TemplateField, "spec", "containers",
+				)
+				image := mapObj[0].(map[string]interface{})["image"]
+				Expect(image).To(BeEquivalentTo(expectedImage))
+			})
+		})
+
+		When("the resource is not in the Kube API", func() {
```

Comment:
Technically it would be more correct to assert that the resource is not in the client before the call to Distribute (or that there arent any resources).